File: kbc/platform/pod/model_usage.py
"""Provider-native Claude stream usage; SDK aggregate defaults are not evidence."""

import uuid
from datetime import datetime, timezone

_FIELDS = ("input_tokens", "output_tokens", "cache_read_input_tokens", "cache_creation_input_tokens")


def _now():
    return datetime.now(timezone.utc).isoformat()


class ClaudeUsageRecorder:
    def __init__(self, session_id, config, observer):
        self.session_id, self.config, self.observer = session_id, config, observer
        self.call = None
        self.seen = set()
        self.raw = {}
        self.terminal = False
        self.response_id = None
        self.response_model = None

    async def observe(self, event, turn_id):
        kind = event.get("type")
        if kind == "message_start":
            await self.finish("incomplete", turn_id)
            model = self.config["model"]
            self.call = {
                "schemaVersion": 1, "callId": str(uuid.uuid4()), "phase": "started",
                "sessionId": self.session_id, "requestId": turn_id, "requestAt": _now(),
                "kind": "agent", "executionRole": "root", "routingAttempt": 1,
                "executorRole": self.config.get("role", "compile"),
                "model": {"configId": "", "name": model.get("name", model["id"]),
                          "sourceKind": "unknown", "sourceId": "", "sourceName": "",
                          "requestedId": model["id"], "runtimeProvider": model["provider"]},
            }
            message = event.get("message") or {}
            self.response_id, self.response_model = message.get("id"), message.get("model")
            self.raw, self.terminal = {}, False
            self._merge(message.get("usage"))
            await self._emit(self.call, turn_id)
        elif kind == "message_delta" and self.call:
            self._merge(event.get("usage"))
        elif kind == "message_stop" and self.call:
            self.terminal = True
            await self.finish("success", turn_id)

    def _merge(self, usage):
        if isinstance(usage, dict):
            for key in _FIELDS:
                if key in usage:
                    # Preserve invalid numeric types as invalid markers, never
                    # coerce them to zero or retain arbitrary provider content.
                    self.raw[key] = usage[key] if type(usage[key]) in (int, float) else "invalid"

    async def _emit(self, observation, turn_id):
        await self.observer.emit("model_usage", {"observation": observation}, turn_id)

    async def finish(self, outcome, turn_id):
        if self.call is None:
            return
        call, self.call = self.call, None
        call = {**call, "phase": "finished", "finishedAt": _now(), "outcome": outcome,
                "usageEvidence": {"protocol": "anthropic", "providerUsagePresent": bool(self.raw),
                                  "finality": "terminal" if self.terminal else "intermediate", "rawUsage": dict(self.raw)}}
        if isinstance(self.response_id, str):
            call["responseId"] = self.response_id[:256]
        if isinstance(self.response_model, str):
            call["responseModel"] = self.response_model[:200]
        if isinstance(self.response_id, str):
            self.seen.add(self.response_id)
        await self._emit(call, turn_id)

    async def record_unobserved(self, message_id, turn_id):
        # A legacy SDK message can establish that a call happened, but its
        # normalized usage is not raw provider evidence.
        if message_id in self.seen or (self.call is not None and message_id == self.response_id):
            return
        model = self.config["model"]
        at = _now()
        observation = {"schemaVersion": 1, "callId": str(uuid.uuid4()), "phase": "finished",
                       "sessionId": self.session_id, "requestId": turn_id, "requestAt": at, "finishedAt": at,
                       "kind": "agent", "executionRole": "root", "routingAttempt": 1, "outcome": "incomplete",
                       "executorRole": self.config.get("role", "compile"),
                       "model": {"configId": "", "name": model.get("name", model["id"]),
                                 "sourceKind": "unknown", "sourceId": "", "sourceName": "",
                                 "requestedId": model["id"], "runtimeProvider": model["provider"]}}
        if message_id:
            self.seen.add(message_id)
        await self._emit(observation, turn_id)

File: src/shared/model-usage.ts
/** Numeric provider evidence is independent of the legacy SDK usage envelope. */
export interface UsageEvidence {
  protocol: string;
  providerUsagePresent: boolean;
  finality: "terminal" | "intermediate" | "missing";
  rawUsage: Record<string, unknown>;
  invalidFields?: string[];
}
export interface UsageIdentity {
  configId: string;
  name: string;
  sourceKind: "api" | "subscription" | "unknown";
  sourceId: string;
  sourceName: string;
}
export interface UsageObservation {
  schemaVersion: 1;
  callId: string;
  phase: "started" | "finished";
  sessionId: string;
  traceId?: string;
  requestId?: string;
  parentCallId?: string;
  executorRole?: string;
  requestAt: string;
  finishedAt?: string;
  kind: "agent" | "aux";
  executionRole: "root" | "internal_subagent" | "delegated";
  routingAttempt: number;
  outcome?: "success" | "error" | "cancelled" | "incomplete";
  finishReason?: string;
  responseModel?: string;
  responseId?: string;
  model: UsageIdentity & { requestedId: string; runtimeProvider: string };
  usageEvidence?: UsageEvidence;
}
export interface UsageHealth {
  collectorId: string;
  sessionId: string;
  pending: number;
  rejected: number;
  dropped: number;
  oldestAt?: string;
  lastSuccessAt?: string;
  capturedAt: string;
}
export interface UsageBatch { observations: UsageObservation[]; health?: UsageHealth }
export interface UsageBatchResponse {
  results: { callId: string; phase: string; status: "accepted" | "duplicate" | "rejected" | "retryable"; reason?: string }[];
}
export interface UsageSink {
  context(): Pick<UsageObservation, "sessionId" | "traceId" | "executionRole" | "requestId" | "parentCallId" | "executorRole">;
  record(observation: UsageObservation): void;
}
